### ISSUE TITLE
[REQUEST] Miri for Rust

### ISSUE THREAD

**Issue by ojeda** [compiler-explorer/compiler-explorer#2563]
Miri (https://github.com/rust-lang/miri) is an experimental interpreter for Rust's mid-level intermediate representation (MIR). It can detect certain classes of undefined behavior, which is very useful when dealing with `unsafe` Rust code.

The Rust playground has support for it under "Tools".

**Comment by narpfel** [comment]
I’ve looked a bit into this: Miri needs a special pre-built standard library which is normally built on-demand when using `cargo miri` and cached locally. This standard library needs to match the exact nightly version of the compiler. The rest of Miri is already installed by `./bin/ce_install ...`.

For CE, the standard lib can be pre-built with a `misc-builder` (I have a draft here https://github.com/compiler-explorer/misc-builder/commit/0922b4ed559440a071dab9e7b83b3bbd17d1ee02), but because the versions must match, this duplicates the whole nightly compiler. Is this okay or is there a way to reuse the normal nightly compiler (run the Miri build after nightly was updated)?

Another approach would be to invoke the build here https://github.com/compiler-explorer/infra/blob/fa186f6a38640cabfa154b99a83bdb6e4e2b2eb4/bin/lib/installable/rust.py#L71-L72. That would avoid duplicating the compiler, but then the Rust could not be installed without compiling the standard lib for each of the dozens of architectures.

**Comment by ojeda** [comment]
Oh, I didn't notice these PRs -- thanks a lot for this!

Cc'ing @RalfJung so that he is aware.

**Comment by RalfJung** [comment]
Oh nice, I had no idea. :)

I can't find it in the UI, but it seems the PR just landed so I guess it'll take a bit before it gets deployed.

**Comment by narpfel** [comment]
> I can't find it in the UI, but it seems the PR just landed so I guess it'll take a bit before it gets deployed.

Yes, it’ll be under “Add tool...” → “Miri nightly” when it’s rolled out: https://github.com/compiler-explorer/compiler-explorer/pull/7718#issuecomment-2922753798

**Comment by ojeda** [comment]
It is live now :)

Example with the arguments support: comparing SB vs. TB: https://godbolt.org/z/PT3E3KeP1

**Comment by RalfJung** [comment]
That's awesome, thanks a lot for doing this @narpfel. :-)

(FWIW, SB vs TB can be done om the playground now -- but none of the other flags is accessible there.)
